### DISCUSSION TITLE
Stage Crear PDF de rutinas entrega pdf en blanco

### DISCUSSION THREAD

**Discussion by AlejandroLoraG** [Cotalker/documentation#651] [4. Bug Reports] [answered]
### Hola a tod@s!! Espero que estén de maravillas y no tan llen@s de bugs <3

_**Este bug está deteniendo un cliente crítico**_

### Descripción del proceso:
Le creamos cotizaciones a los clientes de nuestro cliente (Agrointegral), estas cotizaciones se envían vía mail en forma de pdf adjunto. 
Estos pdf se crean a través de un html que generamos en faas, es decir, se pasa la url del html creado al campo _URL_ del stage Crear PDF (fig 1). Una vez creado el pdf, se envía el mail y se manda en el campo _ADJUNTO_ del stage enviar email (fig 2).

### Descripción del bug:
Al enviar el mail, el pdf llega vacío. Lo curioso es que el bug no ocurre el 100% de las veces, 3 de cada 10 veces (estimado) el pdf llega vacío (fig 3 Y 4).

**_Cabe decir que este bug no es solo con este pdf, en otros flujo del proyecto_**

### Detalles del HTML:
El html se crea con [EJS](https://ejs.co/) (fig 5),  y luego se carga en el response del endpoint con [render](https://www.geeksforgeeks.org/express-js-res-render-function/) de _express.js_ (fig 6). Este endpoint se carga en el campo _URL_ del Creador de PDF (fig 1).

FIG 1
![FIG 1](https://user-images.githubusercontent.com/51095635/149997722-da449df2-6106-4488-ba51-08a80f2fe77b.png)

FIG 2
![FIG 2](https://user-images.githubusercontent.com/51095635/149998123-de86a0ac-ce39-497f-8894-0bd0ef33a0fe.png)

FIG 3
![FIG 3](https://user-images.githubusercontent.com/51095635/149998723-7f2f103f-9893-4a2e-9b74-8778dc7dab25.png)

FIG 4
![FIG 4](https://user-images.githubusercontent.com/51095635/149998819-5cf02cca-7641-48aa-8054-995117576934.png)

FIG 5
![FIG 5](https://user-images.githubusercontent.com/51095635/149998937-01422491-1334-4f44-a835-361b7986a3b6.png)

FIG 6
![FIG 6](https://user-images.githubusercontent.com/51095635/149999045-5c24483f-9099-4b37-8ff2-0766e63aaef9.png)

**Comment by Yiakman** [comment]
Ale, creo que por el lado de la implementación se puede solucionar si se utiliza un stage que llame un endpoint intermedio que gatille un caché del html que se quiere generar en el faas. El problema pasa en que a veces se demora la generación de la data y PBPDF "_saca la foto_" antes de que esté listo el html.
Actualmente está así
```mermaid
sequenceDiagram
BOT->>PBPDF: Gimme this url's pdf
PBPDF->>FAAS: Gimme this url's HTML
loop 
    FAAS->>FAAS: Working on it 🕧
end
FAAS-->>PBPDF: Here you go
PBPDF->>BOT: PDF's URL
```
La solución que propongo es más así
```mermaid
sequenceDiagram
BOT->>FAAS: Load this url template and store it
loop 
    FAAS->>FAAS: Working on it 🕧
end
FAAS->>BOT: Confirmation message
BOT->>PBPDF: Gimme this url's pdf
PBPDF->>FAAS: Gimme this url's HTML
FAAS-->>PBPDF: Here you go, it was in caché
PBPDF->>BOT: PDF's URL
```

**Comment by acedward** [comment, marked as answer]
PBPdf v3 con herurísitca para esperar cargar la página lanzado
